src/Catalog/Helpers/ParallelExecutionUtility.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace NuGet.Services.Metadata.Catalog.Helpers
+{
+    public static class ParallelExecutionUtility
+    {
+        public const int DefaultDegreeOfParallelism = 32;
+
+        /// <summary>
+        /// Runs a task created by <paramref name="taskFactory"/> on a number of threads specified by <paramref name="degreeOfParallelism"/> and waits for all tasks to complete.
+        /// </summary>
+        /// <param name="taskFactory">Creates tasks for each thread to run.</param>
+        /// <param name="degreeOfParallelism">The number of threads to run tasks on.</param>
+        /// <returns>A task that completes when all threads have completed their tasks.</returns>
+        public static Task ProcessInParallel(Func<Task> taskFactory, int degreeOfParallelism = DefaultDegreeOfParallelism)
+        {
+            var bagTakers = new List<Task>();
+            for (int i = 0; i < degreeOfParallelism; i++)
```

Comment:
Is there any reason for not using Parallel.For or Paralle.Foreach instead of ProcessInParallel method?

Reply:
Wasn't aware of it--I'll get rid of this class and use it

Reply:
Decided to keep it because `Parallel.ForEach` is not async.

Reply:
See Stephen Toub's [ForEachAsync](https://blogs.msdn.microsoft.com/pfxteam/2012/03/04/implementing-a-simple-foreachasync) blog posts. You want something similar, but based on Parallel.For... maybe below is cleaner?

```
public static Task ForAsync(int dop, Func<Task> body, CancellationToken ct)
{
    return Task.WhenAll(
        from i in Enumerable.Repeat(body, dop)
        select Task.Run(i, ct)
        );
}
```

Reply:
Maybe just name the class Parallel? So invocation looks like:
```
  Parallel.ForAsync(32, ProcessPackages);
```

Reply:
I really like using `Enumerable.Repeat`, although I've reduced it down to
```
return Task.WhenAll(
    Enumerable
        .Repeat(taskFactory, degreeOfParallelism)
        .Select(f => f()));
```

Reply:
can't make it an extension of `Parallel`, you can't make extension methods on static classes

Reply:
I like `ParallelAsync` however